### ISSUE TITLE
insights: [spike] historical backfill compression may misbehave for time frames prior to the earliest index

### ISSUE THREAD

**Issue by coury-clark** [sourcegraph/sourcegraph#30255]
The historical backfiller uses a [compression mechanism](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+%23%23%23%23+Backfill+compression&patternType=literal) to reduce the number of frames we need to sample based on the commit history of a repository.

Prior to 3.35 we used a fixed width historical size of 12 months for all series, to align with a guarantee that we have indexed at least as much commit history. However, with 3.35 we can now edit the time range to exceed that 12 months of history, which raises some questions about the compression mechanism:
1. Are frames that exceed the oldest index evaluated correctly (ie. not compressed) [in our compression logic](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+func+%28c+*CommitFilter%29+FilterFrames&patternType=literal)?
2. Do we have a useful or good way to determine the oldest frame that has been compressed, and does this differentiate in any meaningful way from the oldest commit in the repository?
3. Should we consider indexing the entire repository history, and if so what would be the performance / storage impact of this? I believe search-product did some early investigations into this space we might be able to learn from.

**Comment by coury-clark** [comment]
I think I just confirmed this is a problem:
`seriesId: 242lOzkcVYjfMwS3LXSwintsZ6N` on k8s-dogfood - for `repo_id: 45849` which corresponds to github.com/DefinitelyTyped/DefinitelyTyped

Series is configured for 9 month intervals.

I just regenerated the series using streaming search and found this repo only got 2 queries: `2021-04-22` and `2022-01-22`. The oldest commit in the repo was `Oct 5 09:39:46 2012 -0700`. Viewing the logs confirms that the compression resulted in this execution plan:
```
[{ 2013-10-22 00:00:00 +0000 UTC [2014-07-22 00:00:00 +0000 UTC 2015-04-22 00:00:00 +0000 UTC 2016-01-22 00:00:00 +0000 UTC 2016-10-22 00:00:00 +0000 UTC 2017-07-22 00:00:00 +0000 UTC 2018-04-22 00:00:00 +0000 UTC 2019-01-22 00:00:00 +0000 UTC 2019-10-22 00:00:00 +0000 UTC 2020-07-22 00:00:00 +0000 UTC]},{bb9f40328b8349b856937b03b2f4d5c5fec789d1 2021-04-22 00:00:00 +0000 UTC []},{21c9f99f8e0f62f0fef7194e52e8782c8c737e0a 2022-01-22 00:00:00 +0000 UTC []}]
```

Unfortunately, there is another bug that caused the first search to be pre-empted as well. As an optimization we preemptively r[emove any queries that would fall before the earliest commit in the repo](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@master/-/blob/internal/vcs/git/commits.go?utm_product_name=GoLand&utm_product_version=GoLand&utm_source=JetBrains-v1.2.1#L420:8). We load the oldest commit; however, [the code that finds the oldest commit in the repo](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@master/-/blob/internal/vcs/git/commits.go?utm_product_name=GoLand&utm_product_version=GoLand&utm_source=JetBrains-v1.2.1#L420:8) can return a commit that isn't the oldest commit, but is just a commit that has no parent commit.

In this case there are 5 commits that can be "no parent", and only one of them would have triggered a search on the first data point.
